staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/matching/matching.go

```diff
@@ -0,0 +1,173 @@
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/predicates/rules"
+)
+
+type MatchCriteria interface {
+	namespace.NamespaceSelectorProvider
+	object.ObjectSelectorProvider
+
+	GetMatchResources() v1alpha1.MatchResources
+}
+
+// Matcher decides if a request matches against matchCriteria
+type Matcher struct {
+	namespaceMatcher *namespace.Matcher
+	objectMatcher    *object.Matcher
+}
+
+// NewMatcher initialize the matcher with dependencies requires
+func NewMatcher(
+	namespaceLister listersv1.NamespaceLister,
+	client kubernetes.Interface,
+) *Matcher {
+	return &Matcher{
+		namespaceMatcher: &namespace.Matcher{
+			NamespaceLister: namespaceLister,
+			Client:          client,
+		},
+		objectMatcher: &object.Matcher{},
+	}
+}
+
+// ValidateInitialization verify if the matcher is ready before use
+func (m *Matcher) ValidateInitialization() error {
+	if err := m.namespaceMatcher.Validate(); err != nil {
+		return fmt.Errorf("namespaceMatcher is not properly setup: %v", err)
+	}
+	return nil
+}
+
+func (m *Matcher) Matches(attr admission.Attributes, o admission.ObjectInterfaces, criteria MatchCriteria, isBinding bool) (bool, error) {
+	matches, err := m.namespaceMatcher.MatchNamespaceSelector(criteria, attr)
+	if err != nil {
+		return false, err
+	}
+	if !matches {
+		return false, nil
+	}
+
+	matches, err = m.objectMatcher.MatchObjectSelector(criteria, attr)
+	if err != nil {
+		return false, err
+	}
+	if !matches {
+		return false, nil
+	}
+
+	matchResources := criteria.GetMatchResources()
+	matchPolicy := matchResources.MatchPolicy
+	if matchesResourceRules(matchResources.ExcludeResourceRules, matchPolicy, attr, o) {
+		return false, nil
+	}
+
+	if isBinding && matchResources.ResourceRules == nil {
+		return true, nil
+	}
+
+	if !matchesResourceRules(matchResources.ResourceRules, matchPolicy, attr, o) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func matchesResourceRules(namedRules []v1alpha1.NamedRuleWithOperations, matchPolicy *v1alpha1.MatchPolicyType, attr admission.Attributes, o admission.ObjectInterfaces) bool {
+	for _, namedRule := range namedRules {
+		rule := v1.RuleWithOperations(namedRule.RuleWithOperations)
+		ruleMatcher := rules.Matcher{
+			Rule: rule,
+			Attr: attr,
+		}
+		if !ruleMatcher.Matches() {
+			continue
+		}
+		// an empty name list always matches
+		if len(namedRule.ResourceNames) == 0 {
+			return true
+		}
+		// TODO: GetName() can return an empty string if the user is relying on
+		// the API server to generate the name... figure out what to do for this edge case
+		name := attr.GetName()
+		for _, matchedName := range namedRule.ResourceNames {
+			if name == matchedName {
+				return true
+			}
+		}
+	}
+
+	// if match policy is undefined or exact, don't perform fuzzy matching
+	// note that defaulting to fuzzy matching is set by the API
+	if matchPolicy == nil || *matchPolicy == v1alpha1.Exact {
+		return false
+	}
+
+	attrWithOverride := &attrWithResourceOverride{Attributes: attr}
```

Comment:
when we match with equivalence, we need to preserve the entry we matched. What we match determines what versioned serialization we need to convert the object to, and what we populate kind/resource in the admissionrequest (requestKind/requestResource come from admission attributes, kind/resource come from what we matched in the matching rule)

Reply:
Addressed with these two commits: https://github.com/cici37/kubernetes/pull/7